File: packages/python/src/opensms/resources/numbers.py
"""The ``numbers`` resource."""

from __future__ import annotations

from typing import List, Optional

from .._serialize import path_id, prune
from ..models import Number, NumberRule, Page
from ._base import Resource


def _rule_body(
    match: str, pattern: Optional[str], action: str, target: str, position: Optional[int]
) -> dict:
    return prune({"match": match, "pattern": pattern, "action": action, "target": target, "position": position})


class Numbers(Resource):
    """Virtual numbers and their inbound routing rules. Everything except
    :meth:`list` and :meth:`available` needs a live key. Accessed as
    ``client.numbers``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[Number]:
        """List numbers assigned to the workspace."""
        return self._page(self._t.request("GET", "/v1/numbers", query={"limit": limit, "cursor": cursor}))

    def available(self, *, country: str, kind: str) -> List[Number]:
        """List numbers available to assign (``kind`` ``long_code``,
        ``short_code`` or ``toll_free``)."""
        return self._t.request("GET", "/v1/numbers/available", query={"country": country, "kind": kind})

    def assign(self, *, country: str, kind: str, idempotency_key: Optional[str] = None) -> Number:
        """Assign a number. Charges the wallet."""
        return self._t.request(
            "POST",
            "/v1/numbers",
            json_body={"country": country, "kind": kind},
            idempotent=True,
            idempotency_key=idempotency_key,
        )

    def release(self, id: str) -> None:
        """Release a number."""
        self._t.request("DELETE", f"/v1/numbers/{path_id(id)}")

    def list_rules(
        self, id: str, *, limit: Optional[int] = None, cursor: Optional[str] = None
    ) -> Page[NumberRule]:
        """List inbound rules for a number."""
        return self._page(
            self._t.request("GET", f"/v1/numbers/{path_id(id)}/rules", query={"limit": limit, "cursor": cursor})
        )

    def create_rule(
        self,
        id: str,
        *,
        match: str,
        action: str,
        target: str,
        pattern: Optional[str] = None,
        position: Optional[int] = None,
        idempotency_key: Optional[str] = None,
    ) -> NumberRule:
        """Add a rule. ``match`` is ``keyword``, ``prefix``, ``regex`` or
        ``any`` (``pattern`` required unless ``any``); ``action`` is
        ``webhook``, ``auto_reply`` or ``forward_email``."""
        return self._t.request(
            "POST",
            f"/v1/numbers/{path_id(id)}/rules",
            json_body=_rule_body(match, pattern, action, target, position),
            idempotent=True,
            idempotency_key=idempotency_key,
        )

    def update_rule(
        self,
        id: str,
        rule_id: str,
        *,
        match: str,
        action: str,
        target: str,
        pattern: Optional[str] = None,
        position: Optional[int] = None,
    ) -> NumberRule:
        """Replace a rule (PUT)."""
        return self._t.request(
            "PUT",
            f"/v1/numbers/{path_id(id)}/rules/{path_id(rule_id, 'rule_id')}",
            json_body=_rule_body(match, pattern, action, target, position),
        )

    def delete_rule(self, id: str, rule_id: str) -> None:
        """Delete a rule."""
        self._t.request("DELETE", f"/v1/numbers/{path_id(id)}/rules/{path_id(rule_id, 'rule_id')}")

File: packages/ruby/lib/opensms/resources/numbers.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +numbers+ resource: virtual numbers and their inbound rules.
    # Everything except +list+ and +available+ needs a live key. Accessed as
    # +client.numbers+.
    class Numbers < Base
      RULE_FIELDS = %i[match pattern action target position].freeze

      # GET /v1/numbers -> Page<Number>.
      def list(params = nil, **kwargs)
        http_page("/v1/numbers", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # GET /v1/numbers/available -> Array<Number>.
      # @param params [Hash] :country, :kind
      def available(params = nil, **kwargs)
        http_get("/v1/numbers/available", Models.build(merge(params, kwargs), %i[country kind]))
      end

      # Assign (buy) a number; charges the wallet. POST /v1/numbers -> 201 Number.
      # @param params [Hash] :country (required), :kind (required)
      def assign(params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), %i[country kind], required: %i[country kind])
        @transport.request(:post, "/v1/numbers", body: body, idempotency_key: idem(idempotency_key))
      end

      # DELETE /v1/numbers/{id} -> 204.
      # @return [nil]
      def release(id)
        http_delete(path("/v1/numbers", id!(id)))
      end

      # GET /v1/numbers/{id}/rules -> Page<Rule>.
      def list_rules(id, params = nil, **kwargs)
        http_page(path("/v1/numbers", id!(id), :rules), Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/numbers/{id}/rules -> 201 Rule.
      # @param rule [Hash] :match, :pattern, :action, :target, :position
      def create_rule(id, rule = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(rule, kwargs), RULE_FIELDS, required: %i[match action target])
        @transport.request(:post, path("/v1/numbers", id!(id), :rules), body: body,
                                                                        idempotency_key: idem(idempotency_key))
      end

      # PUT /v1/numbers/{id}/rules/{rule_id} -> Rule.
      def update_rule(id, rule_id, rule = nil, **kwargs)
        body = Models.build(merge(rule, kwargs), RULE_FIELDS, required: %i[match action target])
        @transport.request(:put, path("/v1/numbers", id!(id), :rules, id!(rule_id, "rule_id")), body: body)
      end

      # DELETE /v1/numbers/{id}/rules/{rule_id} -> 204.
      # @return [nil]
      def delete_rule(id, rule_id)
        http_delete(path("/v1/numbers", id!(id), :rules, id!(rule_id, "rule_id")))
      end
    end
  end
end
